\section{Introduction}

Many fields of computational research are characterized by a growing number of available methods for data analysis. For example, at the time of writing, almost 400 methods are available for analyzing data from single-cell RNA-sequencing experiments \cite{zappia_exploringsinglecellrnaseq_2018}. For experimental researchers and method users, this represents both an opportunity and a challenge, since method choice can significantly affect conclusions.

Benchmarking studies are carried out by computational researchers to compare the performance of different methods, using reference datasets and a range of evaluation criteria. Benchmarks may be performed by authors of new methods to demonstrate performance improvements or other advantages; by independent groups interested in systematically comparing existing methods; or organized as community challenges. 'Neutral' benchmarking studies, i.e., those performed independently of new method development by authors without any perceived bias, and with a focus on the comparison itself, are especially valuable for the research community \cite{boulesteix_necessitydesignstudies_2018,boulesteix_pleaneutralcomparison_2013}.

From our experience conducting benchmarking studies in computational biology, we have learned several key lessons that we aim to synthesize in this review. A number of previous reviews have addressed this topic from a range of perspectives, including: overall commentaries and recommendations on benchmarking design \cite{boulesteix_necessitydesignstudies_2018,peters_puttingbenchmarkstheir_2018,boulesteix_tensimplerules_2015,zheng_benchmarkingcontextsdetails_2017,mangul_systematicbenchmarkingomics_2019,norel_selfassessmenttrapcan_2011,aniba_issuesbioinformaticsbenchmarking_2010}; surveys of design practices followed by existing benchmarks \cite{mangul_systematicbenchmarkingomics_2019}; the importance of neutral benchmarking studies \cite{boulesteix_pleaneutralcomparison_2013}; principles for the design of real-data benchmarking studies \cite{boulesteix_evidencebasedcomputationalstatistics_2017,boulesteix_statisticalframeworkhypothesis_2015} and simulation studies \cite{morris_usingsimulationstudies_2019}; the incorporation of meta-analysis techniques into benchmarking \cite{gardner_identifyingaccuratemetagenome_2019,gardner_metaanalysisbioinformaticssoftware_2017,evangelou_metaanalysismethodsgenomewide_2013,hong_comparisonmetaanalysismethods_2008}; the organization and role of community challenges \cite{boutros_betterbenchmarkingchallengebased_2014,friedberg_tensimplerules_2015}; and discussions on benchmarking design for specific types of methods \cite{vanmechelen_benchmarkingclusteranalysis_2018,angers-loustau_challengesdesigningbenchmark_2018}. More generally, benchmarking may be viewed as a form of meta-research \cite{ioannidis_metaresearchwhyresearch_2018}.

Our aim is to complement previous reviews by providing a summary of essential guidelines for designing, performing, and interpreting benchmarks. While all guidelines are essential for a truly excellent benchmark, some are more fundamental than others. Our target audience consists of computational researchers who are interested in performing a benchmarking study, or who have already begun one. Our review spans the full 'pipeline' of benchmarking, from defining the scope to best practices for reproducibility. This includes crucial questions regarding design and evaluation principles: for example, using rankings according to evaluation metrics to identify a set of high-performing methods, and then highlighting different strengths and tradeoffs among these.

The review is structured as a series of guidelines (Figure \ref{fig:summary_guidelines}), each explained in detail in the following sections. We use examples from computational biology; however, we expect that most arguments apply equally to other fields. We hope that these guidelines will continue the discussion on benchmarking design, as well as assisting computational researchers to design and implement rigorous, informative, and unbiased benchmarking analyses.

\begin{figure}[htb!]
	\centering
	\begin{tcolorbox}
		\begin{enumerate}
			\item Define the purpose and scope of the benchmark.
			\item Include all relevant methods.
			\item Select (or design) representative dataset.
			\item Choose appropriate parameter values and software versions.
			\item Evaluate methods according to key quantitative performance metrics.
			\item Evaluate secondary measures including computational requirements, user-friendliness, installation procedures, and documentation quality.
			\item Interpret results and provide recommendations from both user and method developer perspectives.
			\item Publish results in an accessible format
			\item Design the benchmark to enable future extensions.
			\item Follow reproducible research best practices, by making code and data publicly available.
		\end{enumerate}
  \end{tcolorbox}
	\caption{\textbf{Summary of the guidelines as a set of recommendations.} Each recommendation is discussed in more detail in the corresponding section in the text.}
	\label{fig:summary_guidelines}
\end{figure}

\subsection{Defining the purpose and scope}

The purpose and scope of a benchmark should be clearly defined at the beginning of the study, and will fundamentally guide the design and implementation. In general, we can define three broad types of benchmarking studies: (i) those by method developers, to demonstrate the merits of their approach (e.g., \cite{weber_diffcytdifferentialdiscovery_2019,nowicka_drimseqdirichletmultinomialframework_2016,levine_datadrivenphenotypicdissection_2015,zhou_robustlydetectingdifferential_2014,law_voomprecisionweights_2014}); (ii) neutral studies performed to systematically compare methods for a certain analysis, either conducted directly by an independent group (e.g., \cite{saelens_comparisonsinglecelltrajectory_2019,duo_systematicperformanceevaluation_2018,soneson_biasrobustnessscalability_2018,saelens_comprehensiveevaluationmodule_2018,weber_comparisonclusteringmethods_2016,korthauer_practicalguidemethods_2019,freytag_comparisonclusteringtools_2018,baruzzo_simulationbasedcomprehensivebenchmarking_2017,kanitz_comparativeassessmentmethods_2015,soneson_comparisonmethodsdifferential_2013,rapaport_comprehensiveevaluationdifferential_2013,dillies_comprehensiveevaluationnormalization_2013}) or in collaboration with method authors (e.g., \cite{sage_quantitativeevaluationsoftware_2015}); or (iii) those organized in the form of a community challenge, such as those from the DREAM \cite{weirauch_evaluationmethodsmodeling_2013,costello_communityeffortassess_2014,kuffner_crowdsourcedanalysisclinical_2015,ewing_combiningtumorgenome_2015,hill_inferringcausalmolecular_2016}, FlowCAP \cite{aghaeepour_criticalassessmentautomated_2013,aghaeepour_benchmarkevaluationalgorithms_2016}, CASP \cite{moult_criticalassessmentmethods_2018,moult_criticalassessmentmethods_2016}, CAMI \cite{sczyrba_criticalassessmentmetagenome_2017}, Assemblathon \cite{earl_assemblathoncompetitiveassessment_2011,bradnam_assemblathonevaluatingnovo_2013}, MAQC/SEQC \cite{maqcconsortium_microarrayqualitycontrol_2006,shi_microarrayqualitycontrol_2010,seqc_comprehensiveassessmentrnaseq_2014}, and GA4GH \cite{krusche_bestpracticesbenchmarking_2019} consortia.

A neutral benchmark or community challenge should be as comprehensive as possible, although for any benchmark there will be tradeoffs in terms of available resources. To minimize perceived bias, a research group conducting a neutral benchmark should be approximately equally familiar with all included methods, reflecting typical usage of the methods by independent researchers \cite{boulesteix_pleaneutralcomparison_2013}. Alternatively, the group could include the original method authors, so that each method is evaluated under optimal conditions; methods whose authors decline to take part should be reported. In either case, bias due to focusing attention on particular methods should be avoided -- for example, when tuning parameters or fixing bugs. Strategies to avoid these types of biases, such as the use of blinding, have been previously proposed \cite{boulesteix_evidencebasedcomputationalstatistics_2017}.

By contrast, when introducing a new method, the focus of the benchmark will be on evaluating the relative merits of the new method. This may be sufficiently achieved with a less extensive benchmark, e.g., by comparing against a smaller set of state-of-the-art and baseline methods. However, the benchmark must still be carefully designed to avoid disadvantaging any methods; for example, extensively tuning parameters for the new method while using default parameters for competing methods would result in a biased representation. Some advantages of a new method may fall outside the scope of a benchmark; for example, a new method may enable more flexible analyses than previous methods (e.g., beyond two-group comparisons in differential analyses \cite{weber_diffcytdifferentialdiscovery_2019}).

Finally, results should be summarized in the context of the original purpose of the benchmark. A neutral benchmark or community challenge should provide clear guidelines for method users, and highlight weaknesses in current methods so that these can be addressed by method developers. On the other hand, benchmarks performed to introduce a new method should discuss what the new method offers compared with the current state-of-the-art, such as discoveries that would otherwise not be possible.

\subsection{Selection of methods}

The selection of methods to include in the benchmark will be guided by the purpose and scope of the study. A neutral benchmark should include all available methods for a certain type of analysis. In this case, the publication describing the benchmark will also function as a review of the literature; a summary table describing the methods is a key output (e.g., Figure 2 in \cite{saelens_comparisonsinglecelltrajectory_2019} or Table 1 in \cite{weber_comparisonclusteringmethods_2016}). Alternatively, it may make sense to include only a subset of methods, by defining inclusion criteria: for example, all methods that (i) provide freely available software implementations, (ii) are available for commonly used operating systems, and (iii) can successfully be installed without errors following a reasonable amount of trouble-shooting. Such criteria should be chosen without favoring any methods, and exclusion of any widely used methods should be justified. A useful strategy can be to involve method authors within the process, since they may provide additional details on optimal usage. In addition, community involvement can lead to new collaborations and inspire future method development. However, the overall neutrality and balance of the resulting research team should be maintained. Finally, if the benchmark is organized as a community challenge, the selection of methods will be determined by the participants. In this case, it is important to communicate the initiative widely -- for example, through an established network such as DREAM challenges. However, some authors may choose not to participate; a summary table documenting non-included methods should be provided in this case.

When developing a new method, it is generally sufficient to select a representative subset of existing methods to compare against. For example, this could consist of the current best-performing methods (if known), a simple 'baseline' method, and any methods that are widely used. The selection of competing methods should ensure an accurate and unbiased assessment of the relative merits of the new approach, compared with the current state-of-the-art. In fast-moving fields, for a truly excellent benchmark, method developers should be prepared to update their benchmarks or design them to easily allow extensions as new methods emerge.

\subsection{Selection (or design) of datasets}

The selection of reference datasets is a critical design choice. If suitable publicly accessible datasets cannot be found, they will need to be generated or constructed, either experimentally or by simulation. Including a variety of datasets ensures that methods can be evaluated under a wide range of conditions. In general, reference datasets can be grouped into two main categories: simulated (or synthetic) and real (or experimental).

Simulated data have the advantage that a known true signal (or 'ground truth') can easily be introduced; for example, whether a gene is differentially expressed. Quantitative performance metrics measuring the ability to recover the known truth can then be calculated. However, it is important to demonstrate that simulations accurately reflect relevant properties of real data, by inspecting empirical summaries of both simulated and real datasets (e.g., using automated tools \cite{soneson_unifiedqualityverification_2018}). The set of empirical summaries to use is context-specific; for example, for single-cell RNA-sequencing, dropout profiles and dispersion-mean relationships should be compared \cite{soneson_biasrobustnessscalability_2018}; for DNA methylation, correlation patterns among neighboring CpG sites should be investigated \cite{korthauer_detectionaccuratefalse_2019}; for comparing mapping algorithms, error profiles of the sequencing platforms should be considered \cite{caboche_comparisonmappingalgorithms_2014}. Simplified simulations can also be useful, to evaluate a new method under a basic scenario, or to systematically test aspects such as scalability and stability. However, overly simplistic simulations should be avoided, since these will not provide useful information on performance. A further advantage of simulated data is that it is possible to generate as much data as required; for example, to study variability and draw statistically valid conclusions.

Experimental data often do not contain a ground truth, making it difficult to calculate performance metrics. Instead, methods may be evaluated by comparing them against each other (e.g., overlap between sets of detected differential features \cite{nowicka_drimseqdirichletmultinomialframework_2016}), or against a current widely accepted method or 'gold standard' (e.g., manual gating to define cell populations in high-dimensional cytometry \cite{weber_comparisonclusteringmethods_2016, aghaeepour_criticalassessmentautomated_2013}, or fluorescence in situ hybridization to validate absolute copy number predictions \cite{zheng_benchmarkingcontextsdetails_2017}). In the context of supervised learning, the response variable to be predicted is known in the manually labeled training and test data. However, individual datasets should not be overused, and using the same dataset for both method development and evaluation should be avoided, due to the risk of overfitting and overly optimistic results \cite{grimm_evaluationtoolsused_2015,jelizarow_overoptimismbioinformaticsillustration_2010}. In some cases, it is also possible to design experimental datasets containing a ground truth. Examples include: (i) 'spiking in' synthetic RNA molecules at known relative concentrations \cite{jiang_syntheticspikeinstandards_2011} in RNA-sequencing experiments (e.g., \cite{seqc_comprehensiveassessmentrnaseq_2014,garalde_highlyparalleldirect_2018}), (ii) large-scale validation of gene expression measurements by quantitative polymerase chain reaction (e.g., \cite{seqc_comprehensiveassessmentrnaseq_2014}), (iii) using genes located on sex chromosomes as a proxy for silencing of DNA methylation status (e.g., \cite{law_voomprecisionweights_2014,fang_genomiclandscapehuman_2012}), (iv) using fluorescence-activated cell sorting to sort cells into known subpopulations prior to single-cell RNA-sequencing (e.g., \cite{soneson_biasrobustnessscalability_2018, schaum_singlecelltranscriptomics20_2018, zheng_massivelyparalleldigital_2017}), or (v) mixing different cell lines to create 'pseudo-cells' \cite{tian_benchmarkingsinglecell_2019}. However, it may be difficult to ensure that the ground truth represents an appropriate level of variability -- for example, the variability of spiked-in material, or whether method performance on cell line data is relevant to outbred populations. Alternatively, experimental datasets may be evaluated qualitatively, for example, by judging whether each method can recover previous discoveries, although this strategy relies on the validity of previous results.

A further technique is to design 'semi-simulated' datasets that combine real experimental data with an 'in silico' (i.e., computational) spike-in signal; for example, by combining cells or genes from 'null' (e.g., healthy) samples with a subset of cells or genes from samples expected to contain a true differential signal (examples include \cite{weber_diffcytdifferentialdiscovery_2019,arvaniti_sensitivedetectionrare_2017,rigaill_syntheticdatasets_2018}). This strategy can create datasets with more realistic levels of variability and correlation, together with a ground truth.

Overall, there is no perfect reference dataset, and the selection of appropriate datasets will involve tradeoffs, e.g., regarding the level of complexity. Both simulated and experimental data should not be too 'simple' (e.g., two of the datasets in the FlowCAP-II challenge \cite{aghaeepour_criticalassessmentautomated_2013} gave perfect performance for several algorithms) or too 'difficult' (e.g., for the third dataset in FlowCAP-II, no algorithms performed well); in these situations, it can be impossible to distinguish performance. In some cases, individual datasets have also been found to be unrepresentative, leading to over-optimistic or otherwise biased assessment of methods (e.g., \cite{lowes_bralibasedentatale_2017}). Overall, the key to truly excellent benchmarking is diversity of evaluations, i.e., using a range of metrics and datasets that span the range of those that might be encountered in practice, so that performance estimates can be credibly extrapolated.

\subsection{Parameters and software versions}

Parameter settings can have a crucial impact on performance. Some methods have a large number of parameters, and tuning parameters to optimal values can require significant effort and expertise. For a neutral benchmark, a range of parameter values should ideally be considered for each method, although tradeoffs need to be considered regarding available time and computational resources. Importantly, the selection of parameter values should comply with the neutrality principle, i.e., certain methods should not be favored over others through more extensive parameter tuning.

There are three major strategies for choosing parameters. The first (and simplest) is to use default values for all parameters. Default parameters may be adequate for many methods, although this is difficult to judge in advance. While this strategy may be viewed as too simplistic for some neutral benchmarks, it reflects typical usage. We used default parameters in several neutral benchmarks where we were interested in performance for untrained users \cite{saelens_comparisonsinglecelltrajectory_2019,couronne_randomforestlogistic_2018,schneider_mortalityriskacute_2016}. In addition, for \cite{saelens_comparisonsinglecelltrajectory_2019}, due to the large number of methods and datasets, total runtime was already around a week using 192 processor cores, necessitating judgment in the scope of parameter tuning. The second strategy is to choose parameters based on previous experience or published values. This relies on familiarity with the methods and the literature, reflecting usage by expert users. The third strategy is to use a systematic or automated parameter tuning procedure -- for example, a 'grid search' across ranges of values for multiple parameters or techniques such as cross-validation (e.g., \cite{saelens_comprehensiveevaluationmodule_2018}). The strategies may also be combined, e.g., setting non-critical parameters to default values and performing a grid search for key parameters. Regardless, neutrality should be maintained: comparing methods with the same strategy makes sense, while comparing one method with default parameters against another with extensive tuning makes for an unfair comparison.

For benchmarks performed to introduce a new method, comparing against a single set of optimal parameter values for competing methods is often sufficient; these values may be selected during initial exploratory work or by consulting documentation. However, as outlined above, bias may be introduced by tuning the parameters of the new method more extensively. The parameter selection strategy should be transparently discussed during the interpretation of the results, to avoid the risk of over-optimistic reporting due to expending more 'researcher degrees of freedom' on the new method \cite{boulesteix_tensimplerules_2015,hu_parametertuningkey_2019}.

Software versions can also influence results, especially if updates include major changes to methodology (e.g., \cite{vaquero-garcia_leafcuttervsmajiq_2018}). Final results should generally be based on the latest available versions, which may require re-running some methods if updates become available during the course of a benchmark.

\subsection{Evaluation criteria: key quantitative performance metrics}

Evaluation of methods will rely on one or more quantitative performance metrics (Figure  \ref{fig:fig2_metrics}A). The choice of metric depends on the type of method and data. For example, for classification tasks with a ground truth, metrics include the true positive rate (TPR; sensitivity or recall), false positive rate (FPR; 1 - specificity), and false discovery rate (FDR). For clustering tasks, common metrics include the F1 score, adjusted Rand index, normalized mutual information, precision, and recall; some of these can be calculated at the cluster level as well as averaged (and optionally weighted) across clusters (e.g., these metrics were used to evaluate clustering methods in our own work \cite{duo_systematicperformanceevaluation_2018,weber_comparisonclusteringmethods_2016}  and by others \cite{freytag_comparisonclusteringtools_2018,aghaeepour_criticalassessmentautomated_2013,wiwie_comparingperformancebiomedical_2015}). Several of these metrics can also be compared visually to capture the tradeoff between sensitivity and specificity, e.g., using receiver operating characteristic (ROC) curves (TPR versus FPR), TPR versus FDR curves, or precision-recall (PR) curves (Figure \ref{fig:fig2_metrics}B). For imbalanced datasets, PR curves have been shown to be more informative than ROC curves \cite{saito_precisionrecallplotmore_2015,powers_visualizationtradeoffevaluation_2015}. These visual metrics can also be summarized as a single number, such as area under the ROC or PR curve; examples from our work include \cite{weber_diffcytdifferentialdiscovery_2019,soneson_biasrobustnessscalability_2018}. In addition to the tradeoff between sensitivity and specificity, a method's 'operating point' is important; in particular, whether the threshold used (e.g., 5\% FDR) is calibrated to achieve the specified error rate. We often overlay this onto TPR-FDR curves by filled or open circles (e.g., Figure \ref{fig:fig2_metrics}B, generated using the iCOBRA package \cite{soneson_icobraopenreproducible_2016}); examples from our work include \cite{weber_diffcytdifferentialdiscovery_2019,nowicka_drimseqdirichletmultinomialframework_2016,zhou_robustlydetectingdifferential_2014,soneson_differentialanalysesrnaseq_2016}.

\begin{figure}[htb!]
	\centering
	\includegraphics[width=\hugefigure]{fig/essential_guidelines_figure2} 
	\caption{\textbf{Summary and examples of performance metrics.} \textbf{A:} Schematic overview of classes of frequently used performance metrics, including examples (\textit{boxes outlined in gray}). \textbf{B:} Examples of popular visualizations of quantitative performance metrics for classification methods, using reference datasets with a ground truth. ROC curves (\textit{left}). TPR versus FDR curves (\textit{center}); circles represent observed TPR and FDR at typical FDR thresholds of 1, 5, and 10\%, with filled circles indicating observed FDR lower than or equal to the imposed threshold. PR curves (\textit{right}). Visualizations were generated using iCOBRA R/Bioconductor package \cite{soneson_icobraopenreproducible_2016}. \textit{FDR} false discovery rate, \textit{FPR} false positive rate, \textit{PR} precision-recall, \textit{ROC} receiver operating characteristic, \textit{TPR} true positive rate}
	\label{fig:fig2_metrics}
\end{figure}


For methods with continuous-valued output (e.g., effect sizes or abundance estimates), metrics include the root mean square error, distance measures, Pearson correlation, sum of absolute log-ratios, log-modulus, and cross-entropy. As above, the choice of metric depends on the type of method and data (e.g., \cite{costello_communityeffortassess_2014,lindgreen_evaluationaccuracyspeed_2016}  used correlation, while \cite{moult_criticalassessmentmethods_2016} used root mean square deviation). Further classes of methods include those generating graphs, phylogenetic trees, overlapping clusters, or distributions; these require more complex metrics. In some cases, custom metrics may need to be developed (e.g., we defined new metrics for topologies of developmental trajectories in \cite{saelens_comparisonsinglecelltrajectory_2019}). When designing custom metrics, it is important to assess their reliability across a range of prediction values (e.g., \cite{gurevich_quastqualityassessment_2013,narzisi_comparingnovogenome_2011}). For some metrics, it may also be useful to assess uncertainty, e.g., via confidence intervals. In the context of supervised learning, classification or prediction accuracy can be evaluated by cross-validation, bootstrapping, or on a separate test dataset (e.g., \cite{gardner_identifyingaccuratemetagenome_2019,aghaeepour_benchmarkevaluationalgorithms_2016}). In this case, procedures to split data into training and test sets should be appropriate for the data structure and the prediction task at hand (e.g., leaving out whole samples or chromosomes \cite{schreiber_pitfallmachinelearning_2019}).

Additional metrics that do not rely on a ground truth include measures of stability, stochasticity, and robustness. These measures may be quantified by running methods multiple times using different inputs or subsampled data (e.g., we observed substantial variability in performance for some methods in \cite{soneson_biasrobustnessscalability_2018,weber_comparisonclusteringmethods_2016}). 'Missing values' may occur if a method does not return any values for a certain metric, e.g., due to a failure to converge or other computational issues such as excessive runtime or memory requirements (e.g., \cite{saelens_comparisonsinglecelltrajectory_2019, soneson_biasrobustnessscalability_2018,weber_comparisonclusteringmethods_2016}). Fallback solutions such as imputation may be considered in this case \cite{bischl_benchmarkinglocalclassification_2013}, although these should be transparently reported. For non-deterministic methods (e.g., with random starts or stochastic optimization), variability in performance when using different random seeds or subsampled data should be characterized. Null comparisons can be constructed by randomizing group labels such that datasets do not contain any true signal, which can provide information on error rates (e.g., \cite{weber_diffcytdifferentialdiscovery_2019,zhou_robustlydetectingdifferential_2014,law_voomprecisionweights_2014}). However, these must be designed carefully to avoid confounding by batch or population structure, and to avoid strong within-group batch effects that are not accounted for.

For most benchmarks, multiple metrics will be relevant. Focusing on a single metric can give an incomplete view: methods may not be directly comparable if they are designed for different tasks, and different users may be interested in different aspects of performance. Therefore, a crucial design decision is whether to focus on an overall ranking, e.g., by combining or weighting multiple metrics. In general, it is unlikely that a single method will perform best across all metrics, and performance differences between top-ranked methods for individual metrics can be small. Therefore, a good strategy is to use rankings from multiple metrics to identify a set of consistently high-performing methods, and then highlight the different strengths of these methods. For example, in \cite{weber_comparisonclusteringmethods_2016}, we identified methods that gave good clustering performance, and then highlighted differences in runtimes among these. In several studies, we have presented results in the form of a graphical summary of performance according to multiple criteria (examples include Figure 3 in \cite{saelens_comparisonsinglecelltrajectory_2019} and Figure 5 in \cite{soneson_biasrobustnessscalability_2018} from our work; and Figure 2 in \cite{sage_quantitativeevaluationsoftware_2015} and Figure 6 in \cite{korthauer_practicalguidemethods_2019} from other authors). Identifying methods that consistently underperform can also be useful, to allow readers to avoid these.

\subsection{Evaluation criteria: secondary measures}

In addition to the key quantitative performance metrics, methods should also be evaluated according to secondary measures, including runtime, scalability, and other computational requirements, as well as qualitative aspects such as user-friendliness, installation procedures, code quality, and documentation quality (Figure \ref{fig:fig2_metrics}A). From the user perspective, the final choice of method may involve tradeoffs according to these measures: an adequately performing method may be preferable to a top-performing method that is especially difficult to use.

In our experience, runtimes and scalability can vary enormously between methods (e.g., in our work, runtimes for cytometry clustering algorithms \cite{weber_comparisonclusteringmethods_2016} and metagenome analysis tools \cite{lindgreen_evaluationaccuracyspeed_2016} ranged across multiple orders of magnitude for the same datasets). Similarly, memory and other computational requirements can vary widely. Runtimes and scalability may be investigated systematically, e.g., by varying the number of cells or genes in a single-cell RNA-sequencing dataset \cite{duo_systematicperformanceevaluation_2018,soneson_biasrobustnessscalability_2018}. In many cases, there is a tradeoff between performance and computational requirements. In practice, if computational requirements for a top-performing method are prohibitive, then a different method may be preferred by some users.

User-friendliness, installation procedures, and documentation quality can also be highly variable \cite{mangul_improvingusabilityarchival_2019,mangul_challengesrecommendationsimprove_2019}. Streamlined installation procedures can be ensured by distributing the method via standard package repositories, such as CRAN and Bioconductor for R, or PyPI for Python. Alternative options include GitHub and other code repositories or institutional websites; however, these options do not provide users with the same guarantees regarding reliability and documentation quality. Availability across multiple operating systems and within popular programming languages for data analysis is also important. Availability of graphical user interfaces can further extend accessibility, although graphical-only methods hinder reproducibility and are thus difficult to include in a systematic benchmark.

For many users, freely available and open source software will be preferred, since it is more broadly accessible and can be adapted by experienced users. From the developer perspective, code quality and use of software development best practices, such as unit testing and continuous integration, are also important. Similarly, adherence to commonly used data formats (e.g., GFF/GTF files for genomic features, BAM/SAM files for sequence alignment data, or FCS files for flow or mass cytometry data) greatly improves accessibility and extensibility.

High-quality documentation is critical, including help pages and tutorials. Ideally, all code examples in the documentation should be continually tested, e.g., as Bioconductor does, or through continuous integration.

\subsection{Interpretation, guidelines, and recommendations}

For a truly excellent benchmark, results must be clearly interpreted from the perspective of the intended audience. For method users, results should be summarized in the form of recommendations. An overall ranking of methods (or separate rankings for multiple evaluation criteria) can provide a useful overview. However, as mentioned above, some methods may not be directly comparable (e.g. since they are designed for different tasks), and different users may be interested in different aspects of performance. In addition, it is unlikely that there will be a clear 'winner' across all criteria, and performance differences between top-ranked methods can be small. Therefore, an informative strategy is to use the rankings to identify a set of high-performing methods, and to highlight the different strengths and tradeoffs among these methods. The interpretation may also involve biological or other domain knowledge to establish the scientific relevance of differences in performance. Importantly, neutrality principles should be preserved during the interpretation.

For method developers, the conclusions may include guidelines for possible future development of methods. By assisting method developers to focus their research efforts, high-quality benchmarks can have significant impact on the progress of methodological research.

Limitations of the benchmark should be transparently discussed. For example, in \cite{saelens_comparisonsinglecelltrajectory_2019} we used default parameters for all methods, while in \cite{weber_comparisonclusteringmethods_2016} our datasets relied on manually gated reference cell populations as the ground truth. Without a thorough discussion of limitations, a benchmark runs the risk of misleading readers; in extreme cases, this may even harm the broader research field by guiding research efforts in the wrong directions.

\subsection{Publication and reporting of results}

The publication and reporting strategy should emphasize clarity and accessibility. Visualizations summarizing multiple performance metrics can be highly informative for method users (examples include Figure 3 in \cite{saelens_comparisonsinglecelltrajectory_2019} and Figure 5 in \cite{soneson_biasrobustnessscalability_2018} from our own work; as well as Figure 6 in \cite{korthauer_practicalguidemethods_2019}). Summary tables are also useful as a reference (e.g., \cite{weber_comparisonclusteringmethods_2016,aghaeepour_criticalassessmentautomated_2013}). Additional visualizations, such as flow charts to guide the choice of method for different analyses, are a helpful way to engage the reader (e.g., Figure 5 in \cite{saelens_comparisonsinglecelltrajectory_2019}).

For extensive benchmarks, online resources enable readers to interactively explore the results (examples from our work include \cite{saelens_comparisonsinglecelltrajectory_2019, weber_comparisonclusteringmethods_2016}, which allow users to filter metrics and datasets). Figure 3 displays an example of an interactive website from one of our benchmarks \cite{saelens_comparisonsinglecelltrajectory_2019}, which facilitates exploration of results and assists users with choosing a suitable method. While tradeoffs should be considered in terms of the amount of work required, these efforts are likely to have significant benefit for the community.

\begin{figure}[htb!]
	\centering
	\includegraphics[width=\hugefigure]{fig/essential_guidelines_figure3} 
	\caption{\textbf{Example of an interactive website allowing users to explore the results of one of our benchmarking studies \cite{saelens_comparisonsinglecelltrajectory_2019}.} This website was created using the Shiny framework in R.}
	\label{fig:fig3_interactive}
\end{figure}

In most cases, results will be published in a peer-reviewed article. For a neutral benchmark, the benchmark will be the main focus of the paper. For a benchmark to introduce a new method, the results will form one part of the exposition. We highly recommend publishing a preprint prior to peer review (e.g., on bioRxiv or arXiv) to speed up distribution of results, broaden accessibility, and solicit additional feedback. In particular, direct consultation with method authors can generate highly useful feedback (examples from our work are described in the acknowledgments in \cite{lindgreen_evaluationaccuracyspeed_2016,freyhult_exploringgenomicdark_2007}). Finally, at publication time, considering open access options will further broaden accessibility.

\subsection{Enabling future extensions}

Since new methods are continually emerging \cite{zappia_exploringsinglecellrnaseq_2018}, benchmarks can quickly become out of date. To avoid this, a truly excellent benchmark should be extensible. For example, creating public repositories containing code and data allows other researchers to build on the results to include new methods or datasets, or to try different parameter settings or pre-processing procedures (examples from our work include \cite{saelens_comparisonsinglecelltrajectory_2019,duo_systematicperformanceevaluation_2018,soneson_biasrobustnessscalability_2018,saelens_comprehensiveevaluationmodule_2018,weber_comparisonclusteringmethods_2016}). In addition to raw data and code, it is useful to distribute pre-processed and/or results data (examples include \cite{duo_systematicperformanceevaluation_2018,soneson_biasrobustnessscalability_2018,soneson_icobraopenreproducible_2016} from our work and \cite{wiwie_comparingperformancebiomedical_2015,bokulich_mockrobiotapublicresource_2016,conchuir_webresourcestandardized_2015} from others), especially for computationally intensive benchmarks. This may be combined with an interactive website, where users can upload results from a new method, to be included in an updated comparison either automatically or by the original authors (e.g., \cite{kanitz_comparativeassessmentmethods_2015,cope_benchmarkaffymetrixgenechip_2004,irizarry_comparisonaffymetrixgenechip_2006}). 'Continuous' benchmarks, which are continually updated, are especially convenient (e.g., \cite{barton_nucleotidesgenomeassembler_2014}), but may require significant additional effort.

\subsection{Reproducible research best practices}

Reproducibility of research findings has become an increasing concern in numerous areas of study \cite{ioannidis_whymostpublished_2005}. In computational sciences, reproducibility of code and data analyses has been recognized as a useful 'minimum standard' that enables other researchers to verify analyses \cite{peng_reproducibleresearchcomputational_2011}. Access to code and data has previously enabled method developers to uncover potential errors in published benchmarks due to suboptimal usage of methods \cite{vaquero-garcia_leafcuttervsmajiq_2018,zhou_countbaseddifferentialexpression_2015,zhou_mirnaseqnormalizationcomparisons_2013}. Journal publication policies can play a crucial role in encouraging authors to follow these practices \cite{hofner_reproducibleresearchstatistics_2016}; experience shows that statements that code and data are 'available on request' are often insufficient \cite{boulesteix_makingcomplexprediction_2019}. In the context of benchmarking, code and data availability also provides further benefits: for method users, code repositories serve as a source of annotated code to run methods and build analysis pipelines, while for developers, code repositories can act as a prototype for future method development work.

Parameter values (including random seeds) and software versions should be clearly reported to ensure complete reproducibility. For methods that are run using scripts, these will be recorded within the scripts. In R, the command \texttt{sessionInfo()} gives a complete summary of package versions, the version of R, and the operating system. For methods only available via graphical interfaces, parameters and versions must be recorded manually. Reproducible workflow frameworks, such as the Galaxy platform \cite{afgan_galaxyplatformaccessible_2018}, can also be helpful. A summary table or spreadsheet of parameter values and software versions can be published as supplementary information along with the publication describing the benchmark (e.g., Supporting Information Table S1 in our study \cite{weber_comparisonclusteringmethods_2016}).

Automated workflow management tools and specialized tools for organizing benchmarks provide sophisticated options for setting up benchmarks and creating a reproducible record, including software environments, package versions, and parameter values. Examples include SummarizedBenchmark \cite{kimes_reproduciblereplicablecomparisons_2019}, DataPackageR \cite{finak_datapackagerreproducibledata_2018}, workflowr \cite{blischak_workflowrframeworkreproducible_2019}, and Dynamic Statistical Comparisons \cite{wang_dscdynamicstatistical_2016}. Some tools (e.g., workflowr) also provide streamlined options for publishing results online. In machine learning, OpenML provides a platform to organize and share benchmarks \cite{vanschoren_openmlnetworkedscience_2014}. More general tools for managing computational workflows, including Snakemake \cite{koster_snakemakescalablebioinformatics_2012}, Make, Bioconda \cite{gruning_biocondasustainablecomprehensive_2018}, and conda, can be customized to capture setup information. Containerization tools such as Docker and Singularity may be used to encapsulate a software environment for each method, preserving the package version as well as dependency packages and the operating system, and facilitating distribution of methods to end users (e.g., in our study \cite{saelens_comparisonsinglecelltrajectory_2019}). Best practices from software development are also useful, including unit testing and continuous integration.

Many free online resources are available for sharing code and data, including GitHub and Bitbucket, repositories for specific data types (e.g., ArrayExpress \cite{kolesnikov_arrayexpressupdatesimplifying_2015}, the Gene Expression Omnibus \cite{barrett_ncbigeoarchive_2013}, and FlowRepository \cite{spidlen_flowrepositoryresourceannotated_2012}), and more general data repositories (e.g., figshare, Dryad, Zenodo, Bioconductor ExperimentHub, and Mendeley Data). Customized resources (examples from our work include \cite{soneson_biasrobustnessscalability_2018, soneson_icobraopenreproducible_2016}) can be designed when additional flexibility is needed. Several repositories allow the creation of 'digital object identifiers' (DOIs) for code or data objects. In general, preference should be given to publicly funded repositories, which provide greater guarantees for long-term archival stability \cite{mangul_improvingusabilityarchival_2019,mangul_challengesrecommendationsimprove_2019}.

An extensive literature exists on best practices for reproducible computational research (e.g., \cite{sandve_tensimplerules_2013}). Some practices (e.g., containerization) may involve significant additional work; however, in our experience, almost all efforts in this area prove useful, especially by facilitating later extensions by ourselves or other researchers.

\section{Discussion}
In this review, we have described a set of key principles for designing a high-quality computational benchmark. In our view, elements of all of these principles are essential. However, we have also emphasized that any benchmark will involve tradeoffs, due to limited expertise and resources, and that some principles are less central to the evaluation. Table \ref{tab:essential_principles} provides a summary of examples of key tradeoffs and pitfalls related to benchmarking, along with our judgment of how truly 'essential' each principle is.

\newcommand{\mycola}[2][1]{\multirow{#1}{3.5cm}{#2}}
\newcommand{\mycolb}[2][1]{\multirow{#1}{3.5cm}{#2}}

\begin{table}[ht!]
	\caption{\textbf{Summary of our views regarding 'how essential' each principle is for a truly excellent benchmark, along with examples of key tradeoffs and potential pitfalls relating to each principle}. The higher the number of plus signs, the more central the principle is to the evaluation.} \label{tab:essential_principles}
	\centering
	\footnotesize
	\begin{tabularx}{\textwidth}{|p{3.5cm}p{1cm}p{3.5cm}X|}
		\hline
		\textbf{Principle} & \textbf{How} \mbox{\textbf{essential}}  & \textbf{Tradeoffs} & \textbf{Potential pitfalls} \\ \hline \hline
		\mycola[2]{1. Defining the purpose and score} & +++ & \mycolb[2]{How comprehensive the benchmark should be} & Scope too broad: too much work given available resources \\ 
		& & & Scope too narrow: unrepresentative and possibly misleading results \\ \hline
		{2. Selection of methods} & +++ & {Number of methods to include} & Excluding key methods \\
		\mycola{3. Selection (or design) of datasets} & +++ & \mycolb{Number and types of datasets to include} & Subjectivity in the choice of datasets: e.g., selecting datasets that are unrepresentative of real-world applications \\
		& & & Too few datasets or simulation scenarios Overly simplistic simulations \\ \hline
		\mycola{4. Parameter and software versions} & ++ & \mycolb{Amount of parameter tuning} & Extensive parameter tuning for some methods while using default parameters for others (e.g., competing methods) \\ \hline
		\mycola[3]{5. Evaluation criteria: key quantitative performance metrics} & +++ & \mycolb[3]{Number and types of performance metrics} & Subjectivity in the choice of metrics: e.g., selecting metrics that do not translate to real-world performance \\
		& & & Metrics that give over-optimistic estimates of performance \\
		& & & Methods may not be directly comparable according to individual metrics (e.g., if methods are designed for different tasks) \\ \hline
		\mycola[3]{6. Evaluation criteria: secondary measures} & ++ & \mycolb[3]{Number and types of performance metrics} & Subjectivity of qualitative measures such as user-friendliness, installation procedures, and documentation quality \\
		& & & Subjectivity in relative weighting between multiple metrics \\
		& & & Measures such as runtime and scalability depend on processor speed and memory \\ \hline
		\mycola[2]{7. Interpretation, guidelines, and recommendations} & ++ & \mycolb[2]{Generality versus specificity of recommendations} & Performance differences between top-ranked methods may be minor \\
		& & & Different readers may be interested in different aspects of performance \\ \hline
		\mycola{8. Publication and reporting of results} & + & Amount of resources to dedicate to building online resources & Online resources may not be accessible (or may no longer run) several years later \\ \hline
		\mycola{9. Enabling future extensions} & ++ & \mycolb{Amount of resources to dedicate to ensuring extensibility} & Selection of methods or datasets for future extensions may be unrepresentative (e.g., due to requests from method authors) \\ \hline
		\mycola{10. Reproducible research best practices} & ++ & Amount of resources to dedicate to reproducibility & Some tools may not be compatible or accessible several years later \\ \hline
	\end{tabularx}
\end{table}

A number of potential pitfalls may arise from benchmarking studies (Table \ref{tab:essential_principles}). For example, subjectivity in the choice of datasets or evaluation metrics could bias the results. In particular, a benchmark that relies on unrepresentative data or metrics that do not translate to real-world scenarios may be misleading by showing poor performance for methods that otherwise perform well. This could harm method users, who may select an inappropriate method for their analyses, as well as method developers, who may be discouraged from pursuing promising methodological approaches. In extreme cases, this could negatively affect the research field by influencing the direction of research efforts. A thorough discussion of the limitations of a benchmark can help avoid these issues. Over the longer term, critical evaluations of published benchmarks, so-called meta-benchmarks, will also be informative \cite{boulesteix_evidencebasedcomputationalstatistics_2017,gardner_identifyingaccuratemetagenome_2019,gardner_metaanalysisbioinformaticssoftware_2017}.

Well-designed benchmarking studies provide highly valuable information for users and developers of computational methods, but require careful consideration of a number of important design principles. In this review, we have discussed a series of guidelines for rigorous benchmarking design and implementation, based on our experiences in computational biology. We hope these guidelines will assist computational researchers to design high-quality, informative benchmarks, which will contribute to scientific advances through informed selection of methods by users and targeting of research efforts by developers.