\documentclass[10pt]{book}

%%%%%%%%%%%%%%%%%%%%%%%
%%% IMPORT PACKAGES %%%
%%%%%%%%%%%%%%%%%%%%%%%

% adjust margins (should be done before importing fancyhdr!)
\usepackage[paperheight=24cm,paperwidth=16cm,margin=2.5cm,a4paper]{geometry} %lmargin=2.5cm,rmargin=4cm,tmargin=3cm,bmargin=3cm

% text packages
\usepackage{csquotes}
\usepackage[english]{babel}
\usepackage{setspace}	

% scripting packages
\usepackage{blindtext}

% style packages
\usepackage{lmodern}
\usepackage{fancyhdr}
\usepackage{microtype} % Sublim­i­nal re­fine­ments to­wards ty­po­graph­i­cal per­fec­tion

% used for defining abbreviations, see nomenclature.tex
\usepackage{nomencl}

% figure packages
\usepackage{tikz}
\usetikzlibrary{calc,shapes,arrows,intersections,shadows}
\tikzset{>=latex}
\usepackage{graphicx}
\usepackage[indention=0.5cm,labelsep=colon,font={sf,small},labelfont={bf,sf}]{caption}
\usepackage[indention=0.5cm,font={sf,small},labelfont={bf,sf}]{subcaption}
\usepackage{float}
\usepackage{multirow}
\usepackage{import}
\usepackage{xcolor}
\usepackage{colortbl}

% pseudocode packages
\usepackage{algorithm}
\usepackage{algpseudocode}

% symbols packages
\usepackage{amssymb}
\usepackage{amsmath}

% table packages
\usepackage{tabularx}
\usepackage{tabulary}
\usepackage{booktabs}

% sections in allcaps
\usepackage{titlesec}
%\titleformat{\section} {\Large\bfseries\center}{}{0em}{\hrule height 0.4pt\vspace{2pt}}[\hrule height 0.4pt] \titlespacing*{\section} {0pt}{0pt}{10pt}
%\titleformat{\subsection} {\large\bfseries\center}{}{0em}{\vspace{0pt}} \titlespacing*{\subsection} {0pt}{0pt}{0pt}
%\titleformat{\subsubsection} {\normalsize\bfseries\center}{}{0em}{\vspace{0pt}} \titlespacing*{\subsubsection}  {0pt}{0pt}{0pt}
%\titleformat{\section} [block]{\Large\bfseries\filcenter}{}{0em}{\MakeUppercase}
%\titleformat{\subsection} [block]{\large\bfseries\filcenter}{}{0em}{\MakeUppercase}
%\titleformat{\subsubsection} [block]{\normalsize\bfseries\filcenter}{}{0em}{\MakeUppercase}

% references packages
\usepackage[sorting=none]{biblatex}
\bibliography{library}
\usepackage{bibentry}
\nobibliography*

% link packages
\usepackage{url}
\usepackage{doi}

% add checkmark commands
\usepackage{pifont}
\newcommand{\cmark}{\ding{51}}
\newcommand{\xmark}{\ding{55}}

% boxes around sections
\usepackage{tcolorbox}

% allow to adjust spacing in enumerates and itemizes
\usepackage{enumitem}

% fix named refs for starred sections
\makeatletter
\def\ttl@useclass#1#2{%
	\@ifstar
	{\ttl@labeltrue\@dblarg{#1{#2}}}% {\ttl@labelfalse#1{#2}[]}%
	{\ttl@labeltrue\@dblarg{#1{#2}}}}
\makeatother

%%%%%%%%%%%%%%%%%%%%%
%%% DEFINE HEADER %%%
%%%%%%%%%%%%%%%%%%%%%
\usepackage{fancyhdr}  
\pagestyle{fancy} 
\renewcommand{\chaptermark}[1]{\markboth{#1}{}}
\fancypagestyle{frontmatter}{%    
	\fancyhf{} 	
	\fancyfoot[RO,LE]{\textsf{Page \thepage}} 
	\fancypagestyle{plain}{
		\renewcommand{\headrulewidth}{0pt}
		\fancyhead{}
		\fancyfoot[RO,LE]{\textsf{Page \thepage}}
	} 
}
\fancypagestyle{mainmatter}{%    
	\fancyhf{} 	
	\fancyhead[RO,LE]{\nouppercase{\small \textsf{\leftmark}}}
	\fancyfoot[RO,LE]{\textsf{Page \thepage}} 
}
\setlength{\headheight}{15pt}


\newcounter{letternum}
\newcounter{lettersum}
\setcounter{lettersum}{13}
\newlength{\thumbtopmargin}
\setlength{\thumbtopmargin}{1cm}
\newlength{\thumbbottommargin}
\setlength{\thumbbottommargin}{3cm}
\newlength{\thumbheight}
\pgfmathsetlength{\thumbheight}{%
	(\paperheight-\thumbtopmargin-\thumbbottommargin)/\value{lettersum}}

\newlength{\thumbwidth}
\setlength{\thumbwidth}{1.5cm}

\definecolor{light-gray}{gray}{0.90}
\definecolor{med-gray}{gray}{0.60}
\tikzset{
	thumb/.style={
		%   draw=black,
		fill=light-gray,
		text=black,
		minimum height=\thumbheight, %\thumbheight,
		text width=\thumbwidth,
		outer sep=0pt,%   outer sep=10pt,
		font=\sffamily\Huge,
	}
}
\newcommand{\oddthumb}[1]{%
	\begin{tikzpicture}[remember picture, overlay]
	\node [thumb,text centered,anchor=north east,] at ($%
	(current page.north east)-%
	(0,\thumbtopmargin+\value{letternum}*\thumbheight)%
	$) {#1};
	\end{tikzpicture}
}
\newcommand{\eventhumb}[1]{%
	\begin{tikzpicture}[remember picture, overlay]
	\node [thumb,text centered,anchor=north west,] at ($%
	(current page.north west)-%
	(0,\thumbtopmargin+\value{letternum}*\thumbheight)%
	$) {#1};
	\end{tikzpicture}
}
% create a new command to set a new lettergroup
\newcommand{\lettergroup}[1]{%
	\fancyhead[LO]{\oddthumb{#1}}%
	\fancyhead[RE]{\eventhumb{#1}}%
	\fancypagestyle{chapterstart}{%
		\renewcommand{\headrulewidth}{0pt}
		\renewcommand{\footrulewidth}{0pt}
		\fancyhf{}
		\chead{\oddthumb{#1}}% chapters start only on odd pages
		\fancyfoot[RO,LE]{\textsf{Page \thepage}} %\fancyfoot[RO,LE]{\textsf{Page \thepage}}
	}
	\thispagestyle{chapterstart}
	\stepcounter{letternum}%
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% DEFINE FIGURE HELPERS %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newcommand{\hugefigure}{\textwidth}
\newcommand{\LARGEfigure}{0.8\textwidth} % 1 per col
\newcommand{\LARgefigure}{0.6\textwidth} % 1 per col
\newcommand{\Largefigure}{0.5\textwidth} % 1 per col
\newcommand{\largefigure}{0.37\textwidth} % 2 per col
\newcommand{\normalfigure}{0.3\textwidth} % 3 per col
\newcommand{\smallfigure}{0.24\textwidth} % 3 per col
\newcommand{\Smallfigure}{0.19\textwidth} % 4 per col
\newcommand{\SMALLfigure}{0.15\textwidth} % 5 per col

\makeatletter
\newcommand*{\centerfloat}{%
	\parindent \z@
	\leftskip \z@ \@plus 1fil \@minus \textwidth
	\rightskip\leftskip
	\parfillskip \z@skip}
\makeatother

\newcommand{\picfig}[4]{
\begin{figure}[ht!]
	\centering \centerfloat
	\includegraphics[width=#2]{#1}
	\caption{#4}
	\label{#3}
\end{figure}}

%%%%%%%%%%%%%%%%%%%%%
%%% DEFINE STYLES %%%
%%%%%%%%%%%%%%%%%%%%%

% use Sonny chapter style
\usepackage[Sonny]{fncychap} %Sonny

% set font after chapter sonny style
\usepackage{fontspec}
\setmainfont [Path = fonts/,
UprightFont = *-300,
ItalicFont = *-300-Italic,
BoldFont = *-700,
BoldItalicFont = *-700-Italic
]{MuseoSans}

%% hyperlinks
% load hyperref after fncychap
\usepackage{hyperref}
\usepackage[open,openlevel=1]{bookmark}

\hypersetup{
	colorlinks = true,
%	linkcolor = {blue},
  urlcolor = {blue},
	citecolor = {med-gray}
}

% Set style variables
%\setlength{\parindent}{0cm} 
\renewcommand{\contentsname}{Table of Contents}
\renewcommand{\listfigurename}{List of Figures}
\renewcommand{\listtablename}{List of Tables}
\renewcommand{\appendixname}{}
\renewcommand{\baselinestretch}{1.25}


% si units and chemical stuff
\usepackage{siunitx}
\usepackage[version=3]{mhchem}

% command for adding comments
\newcommand{\mycomment}[1]{\noindent\fbox{\parbox{\textwidth}{\color{red}#1}}}

\begin{document}

% thesis obligatory content: 
% * an abstract in Dutch
% * a list of abbreviations
% * one uniform (compiling) reference list
% * a resume or Curriculum Vitae of the candidate
% * an introductory chapter and additional linking texts between chapters (papers) if the dissertation consists of a compilation of papers
% * acknowledgements of financing institutions
% * apply for an ISBN number for your thesis. This takes 1-2 days.
	
\pagenumbering{alph}

\frontmatter
\pagestyle{empty}

\include{ch0_frontmatter}

% chapters
\mainmatter
\pagestyle{mainmatter}



% TIP FROM WOUTER: Per chapter, write 'update' reflecting on what came out of the paper

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%     CH1: INTRODUCTION     %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Introduction} 
\chaptermark{Introduction} % short title
\label{chap:introduction}
\graphicspath{{./ch1_introduction/}}
\lettergroup{\thechapter}
\begin{quote}
	\textbf{Abstract:} 
	%Recent developments in single-cell transcriptomics have opened new opportunities for studying dynamic processes in immunology in a high-throughput and unbiased manner. Starting from a mixture of cells in different stages of a developmental process, unsupervised trajectory inference algorithms aim to automatically reconstruct the underlying developmental path that cells are following. In this review, we break down the strategies used by this novel class of methods, and organize their components into a common framework, highlighting several practical advantages and disadvantages of the individual methods. We also give an overview of new insights these methods have already provided regarding the wiring and gene regulation of cell differentiation. As the trajectory inference field is still in its infancy, we propose several future developments which will ultimately lead to a global and data-driven way of studying immune cell differentiation.
\end{quote}
\vfill
Partially adapted from:\\
\textbf{Cannoodt, R.}$^*$, Saelens, W.$^*$, and Saeys, Y. Computational methods for trajectory inference from single-cell transcriptomics. \textit{European Journal of Immunology} 46, 11 (2016), 2496--2506. \doi{10.1002/eji.201646347}.\\
Saelens, W.$^*$, \textbf{Cannoodt, R.}$^*$, Todorov, H., and Saeys, Y. A comparison of single-cell trajectory inference methods. \textit{Nature Biotechnology} 37, 5 (2019), 547--554. \doi{10.1038/s41587-019-0071-9}.\\
Todorov, H., \textbf{Cannoodt, R.}, Saelens, W., and Saeys, Y. Network Inference from Single-Cell Transcriptomic Data. \textit{Gene Regulatory Networks} (2019), 235--249. \doi{10.1007/978-1-4939-8882-2_10}. \\
{\footnotesize $^*$ Equal contribution}
\newpage
\include{ch1_introduction/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%        CH2: DYNGEN        %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{dyngen: simulating single cells} 
\chaptermark{dyngen: simulating single cells}
\label{chap:dyngen}
\graphicspath{{./ch2_dyngen/}}
\include{ch2_dyngen/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%     CH3: DYNBENCHMARK     %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{dynbenchmark: A comparison of single-cell trajectory inference methods} 
\chaptermark{dynbenchmark: A comparison of single-cell trajectory inference methods.} % short title
\label{chap:dynbenchmark}
\graphicspath{{./ch3_dynbenchmark/}}
\begin{quote}
	\textbf{Abstract:} Trajectory inference approaches analyze genome-wide omics data from thousands of single cells and computationally infer the order of these cells along developmental trajectories. Although more than 70 trajectory inference tools have already been developed, it is challenging to compare their performance because the input they require and output models they produce vary substantially. Here, we benchmark 45 of these methods on 110 real and 229 synthetic datasets for cellular ordering, topology, scalability and usability. Our results highlight the complementarity of existing tools, and that the choice of method should depend mostly on the dataset dimensions and trajectory topology. Based on these results, we develop a set of guidelines to help users select the best method for their dataset. Our freely available data and evaluation pipeline (\href{https://benchmark.dynverse.org}{benchmark.dynverse.org}) will aid in the development of improved tools designed to analyze increasingly large and complex single-cell datasets.
\end{quote}
\vfill
Adapted from:\\
Saelens, W.$^*$, \textbf{Cannoodt, R.}$^*$, Todorov, H., and Saeys, Y. A comparison of single-cell trajectory inference methods. \textit{Nature Biotechnology} 37, 5 (2019), 547--554. \doi{10.1038/s41587-019-0071-9}.\\
{\footnotesize $^*$ Equal contribution}
\newpage
\include{ch3_dynbenchmark/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%         CH4: DYNO         %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\newpage{\thispagestyle{empty}\cleardoublepage}
%\chapter{dyno: A toolkit for inferring and interpreting trajectories} 
%\chaptermark{dyno: A toolkit for inferring and interpreting trajectories.} % short title
%\label{chap:dyno}
%\begin{quote}
%	\textbf{Abstract:} \blindtext
%\end{quote}
%\vfill
%Adapted from:\\
%\textbf{Cannoodt, R.}, Saelens, W., and Saeys, Y. dyno: A toolkit for inferring and interpreting trajectories. \textit{Journal} vol, issue (2019), page--page. doi.
%\include{ch4_dyno/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%       CH4: SCORPIUS       %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{SCORPIUS: Fast, accurate, and robust single-cell pseudotime} 
\chaptermark{SCORPIUS: Fast, accurate, and robust single-cell pseudotime.} % short title
\label{chap:scorpius}
\begin{quote}
	\textbf{Abstract:} \blindtext
\end{quote}
\vfill
Adapted from:\\
\textbf{Cannoodt, R.}, Saelens, W., Sichien, D., Tavernier, S., Janssens, S. Guilliams, M., Lambrecht, B., De Preter, K., and Saeys, Y. SCORPIUS improves trajectory inference and identifies novel modules in dendritic cell development. \textit{Journal} vol, issue (2019), page--page. \doi{10.1101/079509v2}.
\newpage
\graphicspath{{./ch4_scorpius/}}
\include{ch4_scorpius/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%         CH5: BRED         %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{bred: Inferring single cell regulatory networks} 
\chaptermark{bred: Inferring single cell regulatory networks.} % short title
\label{chap:bred}
\begin{quote}
	\textbf{Abstract:} \blindtext
\end{quote}
\vfill
Adapted from:\\
\textbf{Cannoodt, R.}, Saelens, W., and Saeys, Y. Inferring Single Cell Regulatory Networks. \textit{Journal} vol, issue (2019), page--page. doi.
\newpage
\include{ch5_bred/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%       CH7: INCGRAPH       %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\newpage{\thispagestyle{empty}\cleardoublepage}
%\chapter{incgraph: Optimising regulatory networks} 
%\chaptermark{incgraph: Optimising regulatory networks.} % short title
%\label{chap:incgraph}
%\graphicspath{{./ch7_incgraph/}}
%\begin{quote}
%	\textbf{Abstract:} Graphlets are small network patterns that can be counted in order to characterise the structure of a network (topology).
%	As part of a topology optimisation process, one could use graphlet counts to iteratively modify a network and keep track of the graphlet counts, in order to achieve certain topological properties.
%	Up until now, however, graphlets were not suited as a metric for performing topology optimisation; when millions of minor changes are made to the network structure it becomes computationally intractable to recalculate all the graphlet counts for each of the edge modifications. \\ IncGraph is a method for calculating the differences in graphlet counts with respect to the network in its previous state, which is much more efficient than calculating the graphlet occurrences from scratch at every edge modification made.
%	In comparison to static counting approaches, our findings show IncGraph reduces the execution time by several orders of magnitude.
%	The usefulness of this approach was demonstrated by developing a graphlet-based metric to optimise gene regulatory networks.
%	IncGraph is able to quickly quantify the topological impact of small changes to a network, which opens novel research opportunities to study changes in topologies in evolving or online networks, or develop graphlet-based criteria for topology optimisation. \\
%	IncGraph is freely available as an open-source R package on CRAN (incgraph). The development version is also available on GitHub (\href{http://www.github.com/rcannood/incgraph}{rcannood/incgraph}).
%\end{quote}
%\vfill
%Adapted from:\\
%\textbf{Cannoodt, R.}, Ruyssinck, J., Ramon, J., De Preter, K., and Saeys, Y. IncGraph: Incremental graphlet counting for topology optimisation. \textit{PLOS ONE} 13, 4 (2018), e0195997. \doi{10.1371/journal.pone.0195997}.
%\newpage
%\include{ch7_incgraph/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%      CH6: DISCUSSION      %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Discussion} 
\chaptermark{Discussion.} % short title
\label{chap:discussion}
\graphicspath{{./ch6_discussion/}}
\include{ch6_discussion/content}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%   CH7: Self-assessment    %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Self-assessment in trajectory inference} 
\chaptermark{Self-assessment in trajectory inference.} % short title
\label{chap:selfassessment}
\graphicspath{{./ch7_selfassessment/}}
%\begin{quote}
%	\textbf{Abstract:}
%\end{quote}
%\vfill
%Adapted from:\\
%Weber, L. M., Saelens W., \textbf{Cannoodt, R.}, Soneson, C., Hapfelmeier, A., Gardner, P. P.,
%Boulesteix, A., Saeys, Y., and Robinson, M. D.. 
%Essential guidelines for computational method benchmarking. \textit{Genome Biology} 20, 125 (2019). \doi{10.1186/s13059-019-1738-8}.
\newpage
\include{ch7_selfassessment/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%       CH8: Guidelines       %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Essential guidelines for computational method benchmarking} 
\chaptermark{Essential guidelines for computational method benchmarking.} % short title
\label{chap:guidelines}
\graphicspath{{./ch8_guidelines/}}
\begin{quote}
	\textbf{Abstract:} In computational biology and other sciences, researchers are frequently faced with a choice between several computational methods for performing data analyses. Benchmarking studies aim to rigorously compare the performance of different methods using well-characterized benchmark datasets, to determine the strengths of each method or to provide recommendations regarding suitable choices of methods for an analysis. However, benchmarking studies must be carefully designed and implemented to provide accurate, unbiased, and informative results. Here, we summarize key practical guidelines and recommendations for performing high-quality benchmarking analyses, based on our experiences in computational biology.
\end{quote}
\vfill
Adapted from:\\
Weber, L. M., Saelens W., \textbf{Cannoodt, R.}, Soneson, C., Hapfelmeier, A., Gardner, P. P.,
Boulesteix, A., Saeys, Y., and Robinson, M. D.. 
Essential guidelines for computational method benchmarking. \textit{Genome Biology} 20, 125 (2019). \doi{10.1186/s13059-019-1738-8}.
\newpage
\include{ch8_guidelines/content}




\include{chX_backmatter}

\end{document}
