\PassOptionsToPackage{table}{xcolor}
\documentclass[10pt]{book}

%%%%%%%%%%%%%%%%%%%%%%%
%%% IMPORT PACKAGES %%%
%%%%%%%%%%%%%%%%%%%%%%%

% adjust margins (should be done before importing fancyhdr!)
\usepackage[paperheight=24cm,paperwidth=16cm,margin=2.5cm,a4paper]{geometry} %lmargin=2.5cm,rmargin=4cm,tmargin=3cm,bmargin=3cm

% text packages
\usepackage{csquotes}
\usepackage[english]{babel}
\usepackage{setspace}

% style packages
\usepackage{lmodern}
\usepackage{fancyhdr}
\usepackage{microtype} % Sublim­i­nal re­fine­ments to­wards ty­po­graph­i­cal per­fec­tion
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
\usepackage{parskip} % disable paragraph indent and set paragraph skipping
\setlength{\emergencystretch}{3em}  % prevent overfull lines

% used for defining abbreviations, see nomenclature.tex
\usepackage{nomencl}

% figure packages
\usepackage{tikz}
\usetikzlibrary{calc,shapes,arrows,intersections,shadows}
\tikzset{>=latex}
\usepackage{graphicx}
\usepackage[indention=0.5cm,labelsep=colon,font={sf,small},labelfont={bf,sf}]{caption}
\usepackage[indention=0.5cm,font={sf,small},labelfont={bf,sf}]{subcaption}
\usepackage{float}
\usepackage{multirow}
\usepackage{import}
\usepackage{xcolor}
\usepackage{colortbl}
\usepackage{tcolorbox}
\usepackage[export]{adjustbox}

% pseudocode packages
\usepackage{algorithm}
\usepackage{algpseudocode}
\usepackage{algpseudocode}
\MakeRobust{\Call}

% symbols packages
\usepackage{amssymb}
\usepackage{amsmath}

% table packages
\usepackage{tabularx}
\usepackage{tabulary}
\usepackage{booktabs}

% sections in allcaps
\usepackage{titlesec}
%\titleformat{\section} {\Large\bfseries\center}{}{0em}{\hrule height 0.4pt\vspace{2pt}}[\hrule height 0.4pt] \titlespacing*{\section} {0pt}{0pt}{10pt}
%\titleformat{\subsection} {\large\bfseries\center}{}{0em}{\vspace{0pt}} \titlespacing*{\subsection} {0pt}{0pt}{0pt}
%\titleformat{\subsubsection} {\normalsize\bfseries\center}{}{0em}{\vspace{0pt}} \titlespacing*{\subsubsection}  {0pt}{0pt}{0pt}
%\titleformat{\section} [block]{\Large\bfseries\filcenter}{}{0em}{\MakeUppercase}
%\titleformat{\subsection} [block]{\large\bfseries\filcenter}{}{0em}{\MakeUppercase}
%\titleformat{\subsubsection} [block]{\normalsize\bfseries\filcenter}{}{0em}{\MakeUppercase}

% references packages
\usepackage[sorting=none,refsection=chapter]{biblatex}
\bibliography{library}
\usepackage{bibentry}
\nobibliography*

% link packages
\usepackage{url}
\usepackage{doi}

% add checkmark commands
\usepackage{pifont}
\newcommand{\cmark}{\ding{51}}
\newcommand{\xmark}{\ding{55}}

% allow to adjust spacing in enumerates and itemizes
\usepackage{enumitem}

% fix named refs for starred sections
\makeatletter
\def\ttl@useclass#1#2{%
	\@ifstar
	{\ttl@labeltrue\@dblarg{#1{#2}}}% {\ttl@labelfalse#1{#2}[]}%
	{\ttl@labeltrue\@dblarg{#1{#2}}}}
\makeatother

% si units and chemical stuff
\usepackage{siunitx}
\usepackage[version=3]{mhchem}


%%%%%%%%%%%%%%%%%%%%%%%%%%&%%%
%%%      RMD packages      %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% use upquote for straight quotes in verbatim environments
\usepackage{upquote}

% fancy verbatim
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}

% common commands
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}

% something with figures
\usepackage{grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}

% something with making smaller itemizes
\providecommand{\tightlist}{%
	\setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}

% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

% booktab tables
\usepackage{longtable}
\usepackage{array}
\usepackage{wrapfig}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}

%%%%%%%%%%%%%%%%%%%%%
%%% DEFINE HEADER %%%
%%%%%%%%%%%%%%%%%%%%%
\usepackage{fancyhdr}  
\pagestyle{fancy} 
\renewcommand{\chaptermark}[1]{\markboth{#1}{}}
\fancypagestyle{frontmatter}{%    
	\fancyhf{} 	
	\fancyfoot[RO,LE]{\textsf{Page \thepage}} 
	\fancypagestyle{plain}{
		\renewcommand{\headrulewidth}{0pt}
		\fancyhead{}
		\fancyfoot[RO,LE]{\textsf{Page \thepage}}
	} 
}
\fancypagestyle{mainmatter}{%    
	\fancyhf{} 	
	\fancyhead[RO,LE]{\nouppercase{\small \textsf{\leftmark}}}
	\fancyfoot[RO,LE]{\textsf{Page \thepage}} 
}
\setlength{\headheight}{15pt}

\newcounter{letternum}
\newcounter{lettersum}
\setcounter{lettersum}{13}
\newlength{\thumbtopmargin}
\setlength{\thumbtopmargin}{1cm}
\newlength{\thumbbottommargin}
\setlength{\thumbbottommargin}{3cm}
\newlength{\thumbheight}
\pgfmathsetlength{\thumbheight}{%
	(\paperheight-\thumbtopmargin-\thumbbottommargin)/\value{lettersum}}

\newlength{\thumbwidth}
\setlength{\thumbwidth}{1.5cm}

\definecolor{light-gray}{gray}{0.90}
\definecolor{med-gray}{gray}{0.60}
\tikzset{
	thumb/.style={
		%   draw=black,
		fill=light-gray,
		text=black,
		minimum height=\thumbheight, %\thumbheight,
		text width=\thumbwidth,
		outer sep=0pt,%   outer sep=10pt,
		font=\sffamily\Huge,
	}
}
\newcommand{\oddthumb}[1]{%
	\begin{tikzpicture}[remember picture, overlay]
	\node [thumb,text centered,anchor=north east,] at ($%
	(current page.north east)-%
	(0,\thumbtopmargin+\value{letternum}*\thumbheight)%
	$) {#1};
	\end{tikzpicture}
}
\newcommand{\eventhumb}[1]{%
	\begin{tikzpicture}[remember picture, overlay]
	\node [thumb,text centered,anchor=north west,] at ($%
	(current page.north west)-%
	(0,\thumbtopmargin+\value{letternum}*\thumbheight)%
	$) {#1};
	\end{tikzpicture}
}
% create a new command to set a new lettergroup
\newcommand{\lettergroup}[1]{%
	\fancyhead[LO]{\oddthumb{#1}}%
	\fancyhead[RE]{\eventhumb{#1}}%
	\fancypagestyle{chapterstart}{%
		\renewcommand{\headrulewidth}{0pt}
		\renewcommand{\footrulewidth}{0pt}
		\fancyhf{}
		\chead{\oddthumb{#1}}% chapters start only on odd pages
		\fancyfoot[RO,LE]{\textsf{Page \thepage}} %\fancyfoot[RO,LE]{\textsf{Page \thepage}}
	}
	\thispagestyle{chapterstart}
	\stepcounter{letternum}%
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% DEFINE FIGURE HELPERS %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newcommand{\hugefigure}{\textwidth}
\newcommand{\LARGEfigure}{0.8\textwidth} % 1 per col
\newcommand{\LARgefigure}{0.6\textwidth} % 1 per col
\newcommand{\Largefigure}{0.5\textwidth} % 1 per col
\newcommand{\largefigure}{0.37\textwidth} % 2 per col
\newcommand{\normalfigure}{0.3\textwidth} % 3 per col
\newcommand{\smallfigure}{0.24\textwidth} % 3 per col
\newcommand{\Smallfigure}{0.19\textwidth} % 4 per col
\newcommand{\SMALLfigure}{0.15\textwidth} % 5 per col

\makeatletter
\newcommand*{\centerfloat}{%
	\parindent \z@
	\leftskip \z@ \@plus 1fil \@minus \textwidth
	\rightskip\leftskip
	\parfillskip \z@skip}
\makeatother

\newcommand{\picfig}[4]{
\begin{figure}[ht!]
	\centering \centerfloat
	\includegraphics[width=#2]{#1}
	\caption{#4}
	\label{#3}
\end{figure}}

%%%%%%%%%%%%%%%%%%%%%
%%% DEFINE STYLES %%%
%%%%%%%%%%%%%%%%%%%%%

% use Sonny chapter style
\usepackage[Sonny]{fncychap} %Sonny

% set font after chapter sonny style
\usepackage{fontspec}
\setmainfont [Path = fonts/,
UprightFont = *-300,
ItalicFont = *-300-Italic,
BoldFont = *-700,
BoldItalicFont = *-700-Italic
]{MuseoSans}

%% hyperlinks
% load hyperref after fncychap
\usepackage{hyperref}
\urlstyle{same}  % don't use monospace font for urls
\usepackage[open,openlevel=1]{bookmark}

\hypersetup{
	colorlinks = true,
%	linkcolor = {blue},
  urlcolor = {blue},
	citecolor = {med-gray}
}

% Set style variables
%\setlength{\parindent}{0cm} 
\renewcommand{\contentsname}{Table of Contents}
\renewcommand{\listfigurename}{List of Figures}
\renewcommand{\listtablename}{List of Tables}
\renewcommand{\appendixname}{}
\renewcommand{\baselinestretch}{1.25}

% command for adding comments
\newcommand{\mycomment}[1]{\noindent\fbox{\parbox{\textwidth}{\color{red}#1}}}

\begin{document}

% thesis obligatory content: 
% * an abstract in Dutch
% * a list of abbreviations
% * one uniform (compiling) reference list
% * a resume or Curriculum Vitae of the candidate
% * an introductory chapter and additional linking texts between chapters (papers) if the dissertation consists of a compilation of papers
% * acknowledgements of financing institutions
% * apply for an ISBN number for your thesis. This takes 1-2 days.
	
\pagenumbering{alph}

\frontmatter
\pagestyle{empty}

\include{ch0_frontmatter}

% chapters
\mainmatter
\pagestyle{mainmatter}



% TIP FROM WOUTER: Per chapter, write 'update' reflecting on what came out of the paper

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%     CH1: INTRODUCTION     %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Introduction} 
\chaptermark{Introduction} % short title
\label{chap:introduction}

\lettergroup{\thechapter}

{\large \textbf{Abstract}}
\begin{quote}
	%Recent developments in single-cell transcriptomics have opened new opportunities for studying dynamic processes in a high-throughput and unbiased manner. Starting from a mixture of cells in different stages of a developmental process, unsupervised trajectory inference algorithms aim to automatically reconstruct the underlying developmental path that cells are following. In this review, we break down the strategies used by this novel class of methods, and organize their components into a common framework, highlighting several practical advantages and disadvantages of the individual methods. We also give an overview of new insights these methods have already provided regarding the wiring and gene regulation of cell differentiation. As the trajectory inference field is still in its infancy, we propose several future developments which will ultimately lead to a global and data-driven way of studying immune cell differentiation.
\end{quote}

\vfill
%\newpage
%\graphicspath{{./ch1_introduction/}}
%\include{ch1_introduction/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%        CH2: DYNGEN        %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Paper 1: Benchmarking with \textit{in silico} single cells} 
\chaptermark{Paper 1: Benchmarking with \textit{in silico} single cells.}
\label{chap:dyngen}

{\large \textbf{Abstract}}
\begin{quote}
	\textbf{Purpose:} A common problem of pioneering computational tools is that during their development, there are rarely sufficient datasets available for adequately quantitatively assessing its performance. \\
	\textbf{Results:} We developed \texttt{dyngen}, a multi-modality simulator of single cells. 
	In dyngen, the biomolecular state of an \textit{in silico} cell changes over time according to a set of reactions defined by the cell's gene regulatory network. By simulating single cells in terms of its biomolecular state and reactions, the simulator is easily extendible for adding new modalities or experimental procedures. We demonstrate \texttt{dyngen}'s flexibility by simulating snapshot, time-series and perturbation experiments.\\
	\textbf{Conclusion:} \texttt{dyngen} lays the foundations for benchmarking a wide variety of computational single-cell tools, and can be used to help kick-start the development of future types of analyses.
\end{quote}


\vfill
{\large \textbf{Publication status}}
\begin{quote}
	Manuscript in preparation. \\
	\textbf{Cannoodt R}$^*$, Saelens W$^*$, and Saeys Y.  \\
	{\footnotesize $^*$ Equal contribution}
\end{quote}

\vfill
{\large \textbf{Author contributions}}
\begin{quote}
	$\bullet$ \textbf{R.C.} and W.S. designed the study.\\
	$\bullet$ \textbf{R.C.} and W.S. performed the experiments and analysed the data.\\
	$\bullet$ \textbf{R.C.} and W.S. implemented software packages.\\
	$\bullet$ \textbf{R.C.} wrote the original manuscript.\\
	$\bullet$ \textbf{R.C.} and W.S. reviewed and edited the manuscript. \\
	$\bullet$ Y.S. supervised the project.
\end{quote}

\vfill

%\newpage
%\graphicspath{{./ch2_dyngen/}}
%\include{ch2_dyngen/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%     CH3: DYNBENCHMARK     %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Paper 2: A comparison of single-cell trajectory inference methods} 
\chaptermark{Paper 2: A comparison of single-cell trajectory inference methods.}
\label{chap:dynbenchmark}


{\large \textbf{Abstract}}
\begin{quote}
	\textbf{Purpose:} Trajectory inference approaches analyse genome-wide omics data from thousands of single cells and computationally infer the order of these cells along developmental trajectories. Although more than 75 trajectory inference tools have already been developed, it is challenging to compare their performance because the input they require and output models they produce vary substantially. \\
	\textbf{Results:} Here, we benchmark 45 of these methods on 110 real and 229 synthetic datasets for cellular ordering, topology, scalability and usability. Our results highlight the complementarity of existing tools, and that the choice of method should depend mostly on the dataset dimensions and trajectory topology. We develop a set of guidelines to help users select the best method for their dataset. \\
	\textbf{Conclusion:}  Our freely available data and evaluation pipeline will aid in the development of improved tools designed to analyse increasingly large and complex single-cell datasets.
\end{quote}

\vfill
{\large \textbf{Publication status}}
\begin{quote}
	Published in Nature Biotechnology 37, 5 (2019). \doi{10.1038/s41587-019-0071-9}. \\
	Saelens W$^*$, \textbf{Cannoodt R}$^*$, Todorov H, and Saeys Y. \\
	{\footnotesize $^*$ Equal contribution}
\end{quote}

\vfill
{\large \textbf{Author contributions}}
\begin{quote}
	$\bullet$ \textbf{R.C.}, W.S., H.T. and Y.S. designed the study.\\
	$\bullet$ \textbf{R.C.} and W.S. performed the experiments and analysed the data.\\
	$\bullet$ \textbf{R.C.}, W.S., and H.T. implemented software packages.\\
	$\bullet$ \textbf{R.C.} and W.S. wrote the original manuscript.\\
	$\bullet$ \textbf{R.C.}, W.S., H.T., and Y.S. reviewed and edited the manuscript.\\
	$\bullet$ Y.S. supervised the project.
\end{quote}

\vfill


%\newpage
%\graphicspath{{./ch3_dynbenchmark/}}
%\include{ch3_dynbenchmark/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%         CH4: DYNO         %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Paper 3: A toolkit for inferring and interpreting trajectories} 
\chaptermark{Paper 3: A toolkit for inferring and interpreting trajectories.} % short title
\label{chap:dyno}

{\large \textbf{Abstract}}
\begin{quote}
	\textbf{Purpose:} Recent technological advances allow studying dynamic processes by computationally ordering single-cell omics profiles along a trajectory. Since 2014, at least 75 tools for trajectory inference have been developed. For end-users, however, these methods are difficult to execute and compare, mainly due to high variability in input/output data structures, software requirements, and programming interfaces. \\
	\textbf{Results:} We developed \texttt{dyno}, a toolkit for inferring, visualising and interpreting single-cell trajectories, giving access to more than 50 TI methods. \texttt{dyno} provides downstream analyses such as visualising the trajectory with dot-plots or heatmaps, automated labelling using gene sets, automated inference of the directionality, and detecting genes that are differentially expressed along the trajectory. \\
	\textbf{Conclusion:} By providing a uniform interface for executing and visualising trajectories of 50 methods, \texttt{dyno} allows inferring and comparing trajectories for a much broader audience than previously possible.
\end{quote}

\vfill
{\large \textbf{Publication status}}
\begin{quote}
	Manuscript in preparation. \\
	\textbf{Cannoodt R}$^*$, Saelens W$^*$, and Saeys Y.  \\
	{\footnotesize $^*$ Equal contribution}
\end{quote}

\vfill
{\large \textbf{Author contributions}}
\begin{quote}
	$\bullet$ \textbf{R.C.} and W.S. designed the project.\\
	$\bullet$ \textbf{R.C.} and W.S. implemented software packages.\\
	$\bullet$ \textbf{R.C.} and W.S. wrote the original manuscript.\\
	$\bullet$ Y.S. supervised the project.
\end{quote}

\vfill

%\newpage
%\graphicspath{{./ch4_dyno/}}
%\include{ch4_dyno/content}
%
%\clearpage
%\section{References}
%\printbibliography[heading=none]

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%       CH5: SCORPIUS       %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Paper 4: Fast, accurate, and robust single-cell pseudotime} 
\chaptermark{Paper 4: Fast, accurate, and robust single-cell pseudotime.} % short title
\label{chap:scorpius}


{\large \textbf{Abstract}}
\begin{quote}
	\textbf{Purpose:} 
	Recent advances in single-cell RNA-sequencing allow modelling progression of single cells along a dynamic process of interest as a trajectory. While already more than 75 trajectory inference methods have been developed, comprehensive benchmarking demonstrated that pseudotemporal ordering cells robustly is still a challenging task.\\
	\textbf{Results:} \texttt{SCORPIUS} is a trajectory inference method specialised in pseudotemporally ordering single-cell profiles. Applying \texttt{SCORPIUS} to a large collection of datasets shows that it consistently produces more accurate and robust trajectories in comparison to state-of-the-art trajectory inference methods. We used \texttt{SCORPIUS} to generate novel hypotheses regarding dendritic cell development, which were subsequently validated \textit{in vivo}.\\
	\textbf{Conclusion:} By providing a robust and accurate toolkit for inferring linear trajectories, we aim to make pseudotime analysis as easy-to-use and prevalent as other types of analysis such as differential expression and gene set enrichment.
\end{quote}

\vfill
{\large \textbf{Publication status}}
\begin{quote}
	Published in bioRxiv 079509. \doi{10.1038/s41587-019-0071-9}. \\
	Revised manuscript in preparation. \\
	\textbf{Cannoodt R}, Saelens W, Sichien D, Tavernier S, Janssens S, Guilliams M, Lambrecht B, De Preter K, and Saeys Y. 
\end{quote}

\vfill
{\large \textbf{Author contributions}}
\begin{quote}
	$\bullet$ \textbf{R.C.} and Y.S. designed the study.\\
	$\bullet$ \textbf{R.C.} and W.S. benchmarks and analysed the data.\\
	$\bullet$ S.T. and S.D. generated experimental data.\\
	$\bullet$ \textbf{R.C.} implemented the SCORPIUS software package.\\
	$\bullet$ \textbf{R.C.} wrote the original manuscript.\\
	$\bullet$ \textbf{R.C.}, J.S., G.M, L.B., K.D.P., and Y.S. reviewed and edited the manuscript. \\
	$\bullet$ K.D.P. and Y.S. supervised the project.
\end{quote}

\vfill

%\newpage
%\graphicspath{{./ch5_scorpius/}}
%\include{ch5_scorpius/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%         CH6: BRED         %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Paper 5: Inferring single cell regulatory networks} 
\chaptermark{Paper 5: Inferring single cell regulatory networks.} % short title
\label{chap:bred}

{\large \textbf{Abstract}}
\begin{quote}
	\textbf{Purpose:} \ldots \\
	\textbf{Results:} \ldots \\
	\textbf{Conclusion:} \ldots
\end{quote}

\vfill
{\large \textbf{Publication status}}
\begin{quote}
	Manuscript in preparation. \\
	\textbf{Cannoodt R}, Saeys Y$^*$, and De Preter K$^*$. \\
	{\footnotesize $^*$ Equal contribution}
\end{quote}

\vfill
{\large \textbf{Author contributions}}
\begin{quote}
	$\bullet$ \textbf{R.C.}, K.D.P., and Y.S. designed the study.\\
	$\bullet$ \textbf{R.C.} performed the experiments and analysed the data.\\
	$\bullet$ \textbf{R.C.} implemented the bred software package.\\
	$\bullet$ \textbf{R.C.} wrote the original manuscript.\\
	$\bullet$ \textbf{R.C.}, K.D.P., and Y.S. reviewed and edited the manuscript. \\
	$\bullet$ K.D.P. and Y.S. supervised the project.
\end{quote}

\vfill

%\newpage
%\graphicspath{{./ch6_bred/}}
%\include{ch6_bred/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%       CH7: INCGRAPH       %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Paper 6: Optimising regulatory networks} 
\chaptermark{Paper 6: Optimising regulatory networks.} % short title
\label{chap:incgraph}


{\large \textbf{Abstract}}
\begin{quote}
	\textbf{Purpose:} Graphlets are small network patterns that can be counted in order to characterise the structure of a network (topology). As part of a topology optimisation process, graphlet counts are used to iteratively modify a network in order to achieve certain topological properties.
	Up until now, however, graphlets were not suited as a metric for performing topology optimisation; when millions of minor changes are made to the network structure it becomes computationally intractable to recalculate all the graphlet counts for each of the edge modifications. \\
	\textbf{Results:} IncGraph is calculates the differences in graphlet counts with respect to the network in its previous state.
	IncGraph reduces the execution time by several orders of magnitude when compared to recounting the graphlet counts from scratch at every iteration.
	We demonstrate the usefulness of this approach optimising gene regulatory networks using a graphlet-based fitness score. \\
	\textbf{Conclusion:} IncGraph is able to quickly quantify the topological impact of small changes to a network, which opens novel research opportunities to study changes in topologies in evolving or online networks, or develop graphlet-based criteria for topology optimisation.
\end{quote}

\vfill
{\large \textbf{Publication status}}
\begin{quote}
	Published in PLOS ONE 13, 4 (2018). \doi{10.1371/journal.pone.0195997}. \\
	\textbf{Cannoodt R}, Ruyssinck J, Ramon J, De Preter K, and Saeys Y. 
\end{quote}

\vfill
{\large \textbf{Author contributions}}
\begin{quote}
	$\bullet$ \textbf{R.C.}, Jo.R., Ja.R., and Y.S. designed the study.\\
	$\bullet$ \textbf{R.C.}, Jo.R., and Ja.R. performed the experiments and analysed the data.\\
	$\bullet$ \textbf{R.C.} implemented the incgraph software package.\\
	$\bullet$ \textbf{R.C.} wrote the original manuscript.\\
	$\bullet$ \textbf{R.C.}, Jo.R., Ja.R., K.D.P., and Y.S. reviewed and edited the manuscript. \\
	$\bullet$ Jo.R., K.D.P. and Y.S. supervised the project.
\end{quote}

\vfill

%\newpage
%\graphicspath{{./ch7_incgraph/}}
%\include{ch7_incgraph/content}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%   CH9: Self-assessment    %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Paper 7: Self-assessment in trajectory inference} 
\chaptermark{Paper 7: Self-assessment in trajectory inference.} % short title
\label{chap:selfassessment}


{\large \textbf{Abstract}}
\begin{quote}
	\textbf{Purpose:} \ldots \\
	\textbf{Results:} \ldots \\
	\textbf{Conclusion:} \ldots
\end{quote}

\vfill
{\large \textbf{Publication status}}
\begin{quote}
	Manuscript in preparation. \\
	\textbf{Cannoodt R.}
\end{quote}

\vfill
{\large \textbf{Author contributions}}
\begin{quote}
	$\bullet$ \textbf{R.C.} designed the study.\\
	$\bullet$ \textbf{R.C.} surveyed 75 manuscripts from literature and analysed the data.\\
	$\bullet$ \textbf{R.C.} wrote the original manuscript.
\end{quote}

\vfill

%\newpage
%\graphicspath{{./ch9_selfassessment/}}
%\include{ch9_selfassessment/content}
%\clearpage
%\section{References}
%\printbibliography[heading=none]


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%       CH10: Guidelines    %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Paper 8: Essential guidelines for computational method benchmarking} 
\chaptermark{Paper 8: Essential guidelines for computational method benchmarking.} % short title
\label{chap:guidelines}



{\large \textbf{Abstract}}
\begin{quote}
	\textbf{Purpose:} In computational biology and other sciences, researchers are frequently faced with a choice between several computational methods for performing data analyses. Benchmarking studies aim to rigorously compare the performance of different methods using well-characterized benchmark datasets, to determine the strengths of each method or to provide recommendations regarding suitable choices of methods for an analysis. However, benchmarking studies must be carefully designed and implemented to provide accurate, unbiased, and informative results. \\
	\textbf{Results:}  Here, we summarize key practical guidelines and recommendations for performing high-quality benchmarking analyses, based on our experiences in computational biology. \\
	\textbf{Conclusion:} \ldots
\end{quote}

\vfill
{\large \textbf{Publication status}}
\begin{quote}
	Manuscript published in Genome Biology 20, 125 (2019). \doi{10.1186/s13059-019-1738-8}. \\
	Weber LM, Saelens W, \textbf{Cannoodt R}, Soneson C, Hapfelmeier A, Gardner PP, Boulesteix A, Saeys Y, and Robinson MD 
\end{quote}

\vfill
{\large \textbf{Author contributions}}
\begin{quote}
	$\bullet$ \textbf{R.C.} designed the study.\\
	$\bullet$ \textbf{R.C.} surveyed 75 manuscripts from literature and analysed the data.\\
	$\bullet$ \textbf{R.C.} wrote the original manuscript.\\
	$\bullet$ Y.S. supervised the project.
\end{quote}

\vfill

%\newpage
%\graphicspath{{./ch10_guidelines/}}
%\include{ch10_guidelines/content}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%      CH8: DISCUSSION      %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage{\thispagestyle{empty}\cleardoublepage}
\chapter{Discussion} 
\chaptermark{Discussion.} % short title
\label{chap:discussion}

%\newpage
%\graphicspath{{./ch8_discussion/}}
%\include{ch8_discussion/content}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%        BACKMATTER         %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\include{chX_backmatter}

\end{document}
